compiler/testData/diagnostics/tests/enum/incompatibleEnums_withExtendedTypeCompatibilityChecker.fir.kt

```diff
@@ -0,0 +1,145 @@
+// !LANGUAGE: +ProhibitComparisonOfIncompatibleEnums +ExtendedTypeCompatibilityChecker
+
+enum class E1 {
+    A, B
+}
+
+enum class E2 {
+    A, B
+}
+
+fun foo1(e1: E1, e2: E2) {
+    <!INCOMPATIBLE_ENUM_COMPARISON_ERROR!>e1 == e2<!>
+    <!INCOMPATIBLE_ENUM_COMPARISON_ERROR!>e1 != e2<!>
+
+    <!INCOMPATIBLE_ENUM_COMPARISON_ERROR!>e1 == E2.A<!>
+    <!INCOMPATIBLE_ENUM_COMPARISON_ERROR!>E1.B == e2<!>
+
+    <!INCOMPATIBLE_ENUM_COMPARISON_ERROR!>E1.A == E2.B<!>
+
+    e1 == E1.A
+    E1.A == e1
+    e2 == E2.B
+    E2.B == e2
+}
+
+fun foo2(e1: E1, e2: E2) {
+    when (e1) {
+        E1.A -> {}
+            <!INCOMPATIBLE_ENUM_COMPARISON_ERROR!>E2.A<!> -> {}
+            <!INCOMPATIBLE_ENUM_COMPARISON_ERROR!>E2.B<!> -> {}
+        e1 -> {}
+            <!INCOMPATIBLE_ENUM_COMPARISON_ERROR!>e2<!> -> {}
+        else -> {}
+    }
+}
+
+fun foo3(e1: Enum<E1>, e2: Enum<E2>, e: Enum<*>) {
+    e1 == e
+    <!INCOMPATIBLE_ENUM_COMPARISON_ERROR!>e1 == e2<!>
+
+    e1 == E1.A
+    <!INCOMPATIBLE_ENUM_COMPARISON_ERROR!>e1 == E2.A<!>
+
+    when (e1) {
+        e1 -> {}
+        <!INCOMPATIBLE_ENUM_COMPARISON_ERROR!>e2<!> -> {}
+        e -> {}
+        E1.A -> {}
+            <!INCOMPATIBLE_ENUM_COMPARISON_ERROR!>E2.A<!> -> {}
+        else -> {}
+    }
+
+    when (e) {
+        e -> {}
+        e2 -> {}
+        E1.A -> {}
+        E2.A -> {}
+        else -> {}
+    }
+}
+
+interface MyInterface
+open class MyOpenClass
+
+fun foo4(e1: E1, i: MyInterface, c: MyOpenClass) {
+    <!EQUALITY_NOT_APPLICABLE!>e1 == i<!>
+    <!EQUALITY_NOT_APPLICABLE!>i == e1<!>
+
+    <!EQUALITY_NOT_APPLICABLE!>e1 == c<!>
+    <!EQUALITY_NOT_APPLICABLE!>c == e1<!>
+
+    when (e1) {
+            <!INCOMPATIBLE_TYPES!>i<!> -> {}
+            <!INCOMPATIBLE_TYPES!>c<!> -> {}
```

Comment:
How should we interpret these differences? FE 1.0 reports `INCOMPATIBLE_ENUM_COMPARISON_ERROR` for all six above, which looks like, if either type is an enum type, we should report it? Currently, `FirEqualityOperatorCallChecker` checks both types are enum types.

Reply:
FE1.0 reports `INCOMPATIBLE_ENUM_COMPARISON` does not depend on whether the type is enum or not. For example, in https://github.com/JetBrains/kotlin/blob/8322398133a6cc2287dadbfe817ba635bf047b96/compiler/testData/diagnostics/tests/enum/incompatibleEnumEntryClasses.kt#L10 it doesn't report it even though both types are enum. The implementation in https://github.com/JetBrains/kotlin/blob/cc1cb463b2d53b86ded2e781dea5012b54f62526/compiler/frontend/src/org/jetbrains/kotlin/types/expressions/BasicExpressionTypingVisitor.java#L1451 simply always report `EQUALITY_NOT_APPLICABLE` and fallback to `INCOMPATIBLE_ENUM_COMPARISON` if the first check didn't find anything.

We can tweak `FirEqualityOperatorCallChecker` to report enum if one value is enum if desired. But it would not match FE1.0 behavior either.

Reply:
Actually, let me rearrange the checker to always report `INCOMPATIBLE_TYPE` for `when` and report `INCOMPATIBLE_ENUM_COMPARISON_ERROR` whenever one of the operand in a binary expression is an enum. This seems to result in a behavior that is closer to FE1.0.